Привет, меня зовут Артем. Сегодня я проверю твой проект.
<br> Дальнейшее общение будет происходить на "ты" если это не вызывает никаких проблем.
<br> Желательно реагировать на каждый мой комментарий ('исправил', 'не понятно как исправить ошибку', ...)
<br> Пожалуйста, не удаляй комментарии ревьюера, так как они повышают качество повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:

<div class="alert alert-block alert-success">
<b>Успех:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>

-------------------

Будет очень хорошо, если ты будешь помечать свои действия следующим образом:
<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> ...
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Были внесены следующие изменения ...
</div>

### <font color='orange'>Общее впечатление</font>
* Большое спасибо за проделанную работу. Видно, что приложено много усилий.
* Выводы и рассуждения получились содержательными, их было интересно читать.
* Радует, что тебе удалось разобраться с техникой бутстрапа.
* Отлично, что риск убытков посчитан верно.
* Над этим проектом нужно будет еще немного поработать. Однако, изменения не должны занять много времени.
* С радостью отвечу на твои вопросы, если они есть. Лучше всего их собрать в следующей ячейке. Жду новую версию проекта.

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> Привет, Артем, большое спасибо за проверку проекта! Буду следовать инструкции и постараюсь исправить все ошибки.
</div>

# Выбор локации для скважины

Допустим, вы работаете в добывающей компании «ГлавРосГосНефть». Нужно решить, где бурить новую скважину.

Вам предоставлены пробы нефти в трёх регионах: в каждом 10 000 месторождений, где измерили качество нефти и объём её запасов. Постройте модель машинного обучения, которая поможет определить регион, где добыча принесёт наибольшую прибыль. Проанализируйте возможную прибыль и риски техникой *Bootstrap.*

Шаги для выбора локации:

- В избранном регионе ищут месторождения, для каждого определяют значения признаков;
- Строят модель и оценивают объём запасов;
- Выбирают месторождения с самым высокими оценками значений. Количество месторождений зависит от бюджета компании и стоимости разработки одной скважины;
- Прибыль равна суммарной прибыли отобранных месторождений.

Загружаем библиотеки необходимые для работы.

In [1]:
import pandas as pd
import numpy as np

# импорт библиотек для работы с моделью
# подготовка модели
from sklearn.model_selection import train_test_split 

# модели
from sklearn.linear_model import LinearRegression 

# оценка модели
from sklearn.metrics import mean_squared_error 

<div class="alert alert-block alert-success">
<b>Успех:</b> Отлично, что все импорты собраны в первой ячейке ноутбука! Если у того, кто будет запускать твой ноутбук будут отсутствовать некоторые библиотеки, то он это увидит сразу, а не в процессе!
</div>

## Загрузка, предобработка и подготовка данных

### Загружаем данные

In [2]:
try:
    geo_data_0 = pd.read_csv('/Users/dima/Desktop/python_ya/Project 8/geo_data_0.csv', sep=',', decimal='.') # путь к файлу на компьютере, для работы оффлайн    
except:
    geo_data_0 = pd.read_csv('https://code.s3.yandex.net/datasets/geo_data_0.csv', sep=',', decimal='.') # если путь к файлу на компьютере не доступен, работа с файлом ведется с сайта яндекса
try:
    geo_data_1 = pd.read_csv('/Users/dima/Desktop/python_ya/Project 8/geo_data_1.csv', sep=',', decimal='.')    
except:
    geo_data_1 = pd.read_csv('https://code.s3.yandex.net/datasets/geo_data_1.csv', sep=',', decimal='.')
try:
    geo_data_2 = pd.read_csv('/Users/dima/Desktop/python_ya/Project 8/geo_data_2.csv', sep=',', decimal='.')
except:
    geo_data_2 = pd.read_csv('https://code.s3.yandex.net/datasets/geo_data_2.csv', sep=',', decimal='.')


geo_data_0 данные из Региона 0  
geo_data_1 данные из Региона 1  
geo_data_2 данные из Региона 2

### Изучим данные

In [3]:
geo_data_0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


В первом регионе 10 тыс скважин. В данных нулевых значений нет, типы данных соответствуют условиям задачи.

In [4]:
geo_data_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


Во втором регионе 10 тыс скважин. В данных нулевых значений нет, типы данных соответствуют условиям задачи.

In [5]:
geo_data_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


Во третьем регионе 10 тыс скважин. В данных нулевых значений нет, типы данных соответствуют условиям задачи.

**<span style='color: green;'>Вывод:  
пропуски в данных не обнаружены, типы данных корректные.</span>**

### Проверим данные на дубликаты

In [6]:
geo_data_0.duplicated().sum()

0

In [7]:
geo_data_1.duplicated().sum()

0

In [8]:
geo_data_2.duplicated().sum()

0

**<span style='color: green;'>Вывод:  
явные дубликаты в данных не обнаружены.</span>**

### Подготовка данных для обучения

На первом этапе сделаем индексами в таблицах столбец id. Отсортируем таблицы по id.

In [9]:
geo_data_0.set_index('id', inplace=True)
geo_data_0.sort_index(inplace=True)

In [10]:
geo_data_1.set_index('id', inplace=True)
geo_data_1.sort_index(inplace=True)

In [11]:
geo_data_2.set_index('id', inplace=True)
geo_data_2.sort_index(inplace=True)

In [12]:
geo_data_0.head() #проверка

f0        f1        f2     product
id                                             
006OJ  1.289189 -0.623324  3.642039  142.822141
009eY  1.958091  0.268711  3.702938  112.036815
00AfQ  0.558807 -0.366895 -1.268154   49.659570
00AuZ -0.968699  0.416057  3.553416   73.351787
00BTM -0.908137  0.036964  1.053951   38.664619

In [13]:
geo_data_1.head() #проверка

f0         f1        f2     product
id                                              
0022J -1.757135  -8.432343  1.999313   53.906522
003Gl  2.862304  -2.622592  1.999858   57.085625
003Vx  3.480454 -16.404208  3.000960   80.859783
004LG  5.632612   5.149203  3.004665   80.859783
006oG -4.954731 -11.776300  5.001760  137.945408

In [14]:
geo_data_2.head() #проверка

f0        f1        f2     product
id                                             
009Gl  0.557780 -1.771009  4.646146  130.621072
00AuD  0.896418  0.296979  6.032145  112.891391
00CaL -1.022474 -0.744512  4.243409  132.792852
00Cby -1.183997  2.327663  2.691216   97.174303
00NQN -1.476102  0.455868  2.025839   85.888480

Создадим функцию, которая будет подготавливать данные для обучения модели. Разобьём данные на обучающую и валидационную выборки в соотношении 75:25

In [15]:
def prepare_data(data, TEST_SIZE=0.25, RANDOM_STATE=42):
    X_train, X_val, y_train, y_val = train_test_split(
        data.drop(['product'], axis=1),
        data['product'],
        test_size = TEST_SIZE,
        random_state = RANDOM_STATE,
    )
    return X_train, X_val, y_train, y_val

Подготовим данные из всех трех таблиц.

In [16]:
X_train_0, X_val_0, y_train_0, y_val_0 = prepare_data(geo_data_0, TEST_SIZE=0.25, RANDOM_STATE=42)
X_train_1, X_val_1, y_train_1, y_val_1 = prepare_data(geo_data_1, TEST_SIZE=0.25, RANDOM_STATE=42)
X_train_2, X_val_2, y_train_2, y_val_2 = prepare_data(geo_data_2, TEST_SIZE=0.25, RANDOM_STATE=42)

<div class="alert alert-block alert-success">
<b>Успех:</b> Данные загружены и осмотрены корректно.
</div>

## Обучение и проверка модели

По условию задачи для обучения модели необходимо использовать только линейную регрессию.  
Для обучения модели напишем функцию. Предсказание будем делать на валидационной выборке. Дополнительно функция будет рассчитывать средний запас предсказанного сырья и RMSE модели.

In [17]:
def train_line_model(X_train, X_val, y_train, y_val):
    model = LinearRegression()
    model.fit(X_train, y_train)
    y_pred = model.predict(X_val)
    mean_predict = y_pred.mean() # расчет среднего запаса сырья
    rmse = np.sqrt(mean_squared_error(y_val, y_pred)) # расчет RMSE
    return mean_predict, rmse, y_pred

Обучим модель на данных из трех регионов отдельно.

In [18]:
mean_pred_0, rmse_0, y_pred_0 = train_line_model(X_train_0, X_val_0, y_train_0, y_val_0 )
print(f"Регион 0: Средний запас предсказанного сырья: {mean_pred_0:.2f}, RMSE модели: {rmse_0:.2f}")

mean_pred_1, rmse_1, y_pred_1 = train_line_model(X_train_1, X_val_1, y_train_1, y_val_1)
print(f"Регион 1: Средний запас предсказанного сырья: {mean_pred_1:.2f}, RMSE модели: {rmse_1:.2f}")

mean_pred_2, rmse_2, y_pred_2 = train_line_model(X_train_2, X_val_2, y_train_2, y_val_2)
print(f"Регион 2: Средний запас предсказанного сырья: {mean_pred_2:.2f}, RMSE модели: {rmse_2:.2f}")

Регион 0: Средний запас предсказанного сырья: 92.34, RMSE модели: 37.75
Регион 1: Средний запас предсказанного сырья: 68.70, RMSE модели: 0.89
Регион 2: Средний запас предсказанного сырья: 95.01, RMSE модели: 40.08


**<span style='color: green;'>Вывод:  
Модель предсказывает высокий средний запас сырья для региона 0 и 2, но при этом ошибка модели существенно больше. Для региона 1 модель предсказывает меньший запас сырья, но предсказание намного точнее.</span>**

Сохраним полученные предсказанные данные и правильные ответы. Используем только валидационные выборки.

In [19]:
predict_and_actual_0 = pd.DataFrame({'predicted': y_pred_0, 'actual': y_val_0})
predict_and_actual_1 = pd.DataFrame({'predicted': y_pred_1, 'actual': y_val_1})
predict_and_actual_2 = pd.DataFrame({'predicted': y_pred_2, 'actual': y_val_2})

<div class="alert alert-block alert-success">
<b>Успех:</b> Модели обучены корректно. Радует, что ты активно используешь функции. Это помогает избежать дублирования кода, а значит уменьшает вероятность совершить ошибку/опечатку!
</div>

## Подготовка к расчёту прибыли

Определим основные переменные необходимые для расчета прибыли.

In [20]:
BUDGET=10000000000 # бюджет на разработку скважин
INCOME=450000 # доход от с единицы продукта (с тысячи баррелей)
TOTAL_WELL=500 # количество исследуемых точек
WELL_FOR_DEVELOP=200 # лучшие для разработки 

Расчет достаточного объема сырья для безубыточной разработки новой скважены.  
Для этого выделяемый бюджет на разработку (10 млд) резделим на произведение дохода с одной единицы сырья на количество скважин идущих в разработку.

In [21]:
breakeven = BUDGET / (INCOME * WELL_FOR_DEVELOP)
print(f'Безубыточный объём сырья: {breakeven:.2f} тыс. баррелей')

Безубыточный объём сырья: 111.11 тыс. баррелей


**<span style='color: green;'>Вывод:  
Безубыточный объем сырья составляет 111.11 тыс баррелей. Все три модели предсказали меньшие значения, чем нужно. Но разброс в предсказании модели 0 и 2 достаточно большой 38 и 40 тыс баррелей соответственно. В значения разброса обоих предсказаний моделей 0 и 2 попадает значение безубыточной разработки. Таким образом, необходимо пересмотреть две модели (по данным 0 и 2), для более точного предсказания. Регион 1 не подходит под заданные условия исходя из предсказаний модели. </span>**

<div class="alert alert-block alert-info">

<b>Совет: </b> В названиях константных переменных лучше использовать только БОЛЬШИЕ буквы. Это соглашение между программистами: https://pythonworld.ru/osnovy/pep-8-rukovodstvo-po-napisaniyu-koda-na-python.html#id29 .
</div>

<div class="alert alert-block alert-success">
<b>Успех:</b> Точка безубыточности найдена верно, сравнение проведено!
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> Упустил момент про константы, теперь буду знать. Спасибо за ссылку, обязательно изучу!
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Были внесены следующие изменения: все константы переименовал. Соответственно далее в коде название были изменены.
</div>

## Расчёт прибыли и рисков 

### Расчёт прибыли

Составим функцию для расчета прибыли по 200 скважинам с максимально предсказанными значениями добычи. Для этого просуммируем объем сырья 200 выбранных скважин (значения из актуальных данных, не предсказанных) и рассчитаем прибыль от полученного объема.

In [22]:
def calculation_model_profit(data, well_for_develop=WELL_FOR_DEVELOP):
    wells = data.sort_values(by='predicted', ascending=False).head(well_for_develop) # отсортируем данные по пердсказанным значениям и выберем первые 200 значений
    total_volume = wells['actual'].sum() # общий объем дабычи по выбранным скважинам
    profit = (total_volume * INCOME) - BUDGET # расчет прибыли
    return profit

Рассчитаем прибыль для каждого региона.

In [23]:
profit_0 = calculation_model_profit(predict_and_actual_0)
profit_1 = calculation_model_profit(predict_and_actual_1)
profit_2 = calculation_model_profit(predict_and_actual_2)

print(f'Прибыль в регионе 0: {profit_0 / 1e9:.2f} млрд рублей')
print(f'Прибыль в регионе 1: {profit_1 / 1e9:.2f} млрд рублей')
print(f'Прибыль в регионе 2: {profit_2 / 1e9:.2f} млрд рублей')


Прибыль в регионе 0: 3.34 млрд рублей
Прибыль в регионе 1: 2.42 млрд рублей
Прибыль в регионе 2: 2.46 млрд рублей


**<span style='color: green;'>Вывод:  
Все три региона принесут прибыль при разработке 200 лучших скважин, которые выбрала модель.</span>**

### Оценка рисков

Применим технику bootstrap со значением выборки 1000 для того, чтобы найти среднее прибыли, 95%-й доверительный интервал и риск убытков.

#### Регион #0

In [24]:
state = np.random.RandomState(12345)
profits_0 = [] #список для хранения прибыли
for i in range(1000):
    # Создаем подвыборки для бутстрепа по 500 элементов 
    sample_0 = predict_and_actual_0.sample(n=TOTAL_WELL, replace=True, random_state=state)
    # Выбираем 200 скважин с max предсказ значением
    wells_0 = sample_0.sort_values(by='predicted', ascending=False).head(200)
    # Считаем реальны объем для 200 скважин с max предсказ значениями
    total_volume_0 = wells_0['actual'].sum()
    # Рассчитываем прибыль для этого объема
    profit_0 = (total_volume_0 *INCOME) - BUDGET
    # Добавляем полученную прибыль в список
    profits_0.append(profit_0)
# Преобразуем список в серию
profits_0 = pd.Series(profits_0)
# Средняя прибыль
mean_profit_0 = profit_0.mean()
# Рассчитаем 2.5 и 97.5 процентили
lower_bound_0 = profits_0.quantile(0.025)
upper_bound_0 = profits_0.quantile(0.975)
# Расчет доли случаев, где прибыль отрицательна (риски)
risk_0 = (profits_0 < 0).mean()
# Перевод в проценты
risk_0_percent = risk_0*100
# Вывод результатов
print(f'Средняя прибыль для региона 0: {mean_profit_0 / 1e9:.2f} млрд рублей')
print(f'95%-й доверительный интервал для региона 0: от {lower_bound_0 / 1e9:.2f} до {upper_bound_0 / 1e9:.2f} млрд рублей')
print(f'Риск убытков для региона 0: {risk_0_percent:.2f}%')

Средняя прибыль для региона 0: 0.49 млрд рублей
95%-й доверительный интервал для региона 0: от -0.05 до 0.92 млрд рублей
Риск убытков для региона 0: 3.70%


<div class="alert alert-block alert-success">
<b>Успех:</b> Отлично, что для подсчета доверительного интервала использовался именно метод квантилей!
</div>

<div class="alert alert-block alert-success">
<b>Успех:</b> Риск найден очень элегантно.
</div>

<div class="alert alert-block alert-danger">

<b>Ошибка:</b> А почему верхняя граница – это 0.974, а не 0.975?
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> Это была опечатка. А дальше получилось как раз то, о чем ты написал в комментарии, я скопировал код, сохранив эту ошибку. Надо было все-таки не лениться и тоже составить функцию для bootstrap.
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Были внесены следующие изменения: 0.974 исправлено на 0.975. Для расчета 0.95 % доверительного интервала, нужно взять 0.975 и 0.025 квантили. Изменения внесены в расчеты трех регионов.  
В самом конце проекта добавил функцию для расчета bootstrap. Проверил на Регионе 2, вроде работает корректно. 
</div>

#### Регион #1

In [25]:
state = np.random.RandomState(12345)
profits_1 = [] #список для хранения прибыли
for i in range(1000):
    # Создаем подвыборки для бутстрепа по 500 элементов 
    sample_1 = predict_and_actual_1.sample(n=TOTAL_WELL, replace=True, random_state=state)
    # Выбираем 200 скважин с max предсказ значением
    wells_1 = sample_1.sort_values(by='predicted', ascending=False).head(200)
    # Считаем реальны объем для 200 скважин с max предсказ значениями
    total_volume_1 = wells_1['actual'].sum()
    # Рассчитываем прибыль для этого объема
    profit_1 = (total_volume_1 *INCOME) - BUDGET
    # Добавляем полученную прибыль в список
    profits_1.append(profit_1)
# Преобразуем список в серию
profits_1 = pd.Series(profits_1)
# Средняя прибыль
mean_profit_1 = profit_1.mean()
# Рассчитаем 2.5 и 97.5 процентили
lower_bound_1 = profits_1.quantile(0.025)
upper_bound_1 = profits_1.quantile(0.975)
# Расчет доли случаев, где прибыль отрицательна (риски)
risk_1 = (profits_1 < 0).mean()
# Перевод в проценты
risk_1_percent = risk_1*100
# Вывод результатов
print(f'Средняя прибыль для региона 1: {mean_profit_1 / 1e9:.2f} млрд рублей')
print(f'95%-й доверительный интервал для региона 1: от {lower_bound_1 / 1e9:.2f} до {upper_bound_1 / 1e9:.2f} млрд рублей')
print(f'Риск убытков для региона 1: {risk_1_percent:.2f}%')

Средняя прибыль для региона 1: 0.29 млрд рублей
95%-й доверительный интервал для региона 1: от 0.05 до 0.84 млрд рублей
Риск убытков для региона 1: 1.50%


#### Регион #2

In [26]:
state = np.random.RandomState(12345)
profits_2 = [] #список для хранения прибыли
for i in range(1000):
    # Создаем подвыборки для бутстрепа по 500 элементов 
    sample_2 = predict_and_actual_2.sample(n=TOTAL_WELL, replace=True, random_state=state)
    # Выбираем 200 скважин с max предсказ значением
    wells_2 = sample_2.sort_values(by='predicted', ascending=False).head(200)
    # Считаем реальны объем для 200 скважин с max предсказ значениями
    total_volume_2 = wells_2['actual'].sum()
    # Рассчитываем прибыль для этого объема
    profit_2 = (total_volume_2 *INCOME) - BUDGET
    # Добавляем полученную прибыль в список
    profits_2.append(profit_2)
# Преобразуем список в серию
profits_2 = pd.Series(profits_2)
# Средняя прибыль
mean_profit_2 = profit_2.mean()
# Рассчитаем 2.5 и 97.5 процентили
lower_bound_2 = profits_2.quantile(0.025)
upper_bound_2 = profits_2.quantile(0.975)
# Расчет доли случаев, где прибыль отрицательна (риски)
risk_2 = (profits_2 < 0).mean()
# Перевод в проценты
risk_2_percent = risk_2*100

# Вывод результатов
print(f'Средняя прибыль для региона 2: {mean_profit_2 / 1e9:.2f} млрд рублей')
print(f'95%-й доверительный интервал для региона 2: от {lower_bound_2 / 1e9:.2f} до {upper_bound_2 / 1e9:.2f} млрд рублей')
print(f'Риск убытков для региона 2: {risk_2_percent:.2f}%')

Средняя прибыль для региона 2: 0.48 млрд рублей
95%-й доверительный интервал для региона 2: от -0.14 до 0.88 млрд рублей
Риск убытков для региона 2: 7.20%


Соберем все результаты в сводную таблицу.

In [27]:
data = {
    'Регион': ['Регион 0', 'Регион 1', 'Регион 2'],
    'Средняя прибыль (млрд руб)': [mean_profit_0 / 1e9, mean_profit_1 / 1e9, mean_profit_2 / 1e9],
    'Нижняя граница 95% (млрд руб)': [lower_bound_0 / 1e9, lower_bound_1 / 1e9, lower_bound_2 / 1e9],
    'Верхняя граница 95% (млрд руб)': [upper_bound_0 / 1e9, upper_bound_1 / 1e9, upper_bound_2 / 1e9],
    'Риск убытков (%)': [risk_0_percent, risk_1_percent, risk_2_percent]
}
summary_table = pd.DataFrame(data)

summary_table

Регион  Средняя прибыль (млрд руб)  Нижняя граница 95% (млрд руб)  \
0  Регион 0                    0.486727                      -0.051565   
1  Регион 1                    0.292206                       0.048941   
2  Регион 2                    0.482384                      -0.136951   

   Верхняя граница 95% (млрд руб)  Риск убытков (%)  
0                        0.921907               3.7  
1                        0.844541               1.5  
2                        0.883963               7.2

**<span style='color: green;'>Вывод:  
Bootsrap анализ показал, что наименьший риск получить убытки будет при разработке скважин из Региона 1. Но при меньших рисках со скважин данного региона будет меньше средняя прибыль. Но с вероятность 95% данная прибуль будет положительной.   
Таким образом, оптимальным регионом для начала разработок будет Регион 1. 
Вторым можно выбрать Регион 0, но вероятность риска убытка в этом регионе выше заданного по условию.   
Регион 2 выбирать не нужно, при той же прибыли, что и в Регионе 0, риск убытков выше.</span>**

<div class="alert alert-block alert-success">
<b>Успех:</b> С выбором региона согласен.
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> Еще раз большое спасибо за проверку проекта и советы!
</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные подготовлены
- [x]  Выполнен шаг 2: модели обучены и проверены
    - [x]  Данные корректно разбиты на обучающую и валидационную выборки
    - [x]  Модели обучены, предсказания сделаны
    - [x]  Предсказания и правильные ответы на валидационной выборке сохранены
    - [x]  На экране напечатаны результаты
    - [x]  Сделаны выводы
- [x]  Выполнен шаг 3: проведена подготовка к расчёту прибыли
    - [x]  Для всех ключевых значений созданы константы Python
    - [x]  Посчитано минимальное среднее количество продукта в месторождениях региона, достаточное для разработки
    - [x]  По предыдущему пункту сделаны выводы
    - [x]  Написана функция расчёта прибыли
- [x]  Выполнен шаг 4: посчитаны риски и прибыль
    - [x]  Проведена процедура *Bootstrap*
    - [x]  Все параметры бутстрепа соответствуют условию
    - [x]  Найдены все нужные величины
    - [x]  Предложен регион для разработки месторождения
    - [x]  Выбор региона обоснован

In [28]:
def bootstrap_for_well (data_well, n):
    state = np.random.RandomState(12345)
    profits = [] #список для хранения прибыли
    for i in range(1000):
        # Создаем подвыборки для бутстрепа по 500 элементов 
        sample = data_well.sample(n, replace=True, random_state=state)
        # Выбираем 200 скважин с max предсказ значением
        wells = sample.sort_values(by='predicted', ascending=False).head(200)
        # Считаем реальны объем для 200 скважин с max предсказ значениями
        total_volume = wells['actual'].sum()
        # Рассчитываем прибыль для этого объема
        profit = (total_volume *INCOME) - BUDGET
        # Добавляем полученную прибыль в список
        profits.append(profit)
    # Преобразуем список в серию
    profits = pd.Series(profits)
    # Средняя прибыль
    mean_profit = profit.mean()
    # Рассчитаем 2.5 и 97.5 процентили
    lower_bound = profits.quantile(0.025)
    upper_bound = profits.quantile(0.975)
    # Расчет доли случаев, где прибыль отрицательна (риски)
    risk = (profits < 0).mean()
    # Перевод в проценты
    risk_percent = risk*100

    # Вывод результатов
    return mean_profit, lower_bound, upper_bound, risk_percent

In [29]:
mean_profit,lower_bound, upper_bound, risk_percent = bootstrap_for_well(predict_and_actual_2, TOTAL_WELL)

In [30]:
print(f'Средняя прибыль для региона 2: {mean_profit / 1e9:.2f} млрд рублей')
print(f'95%-й доверительный интервал для региона 2: от {lower_bound / 1e9:.2f} до {upper_bound / 1e9:.2f} млрд рублей')
print(f'Риск убытков для региона 2: {risk_percent:.2f}%')

Средняя прибыль для региона 2: 0.48 млрд рублей
95%-й доверительный интервал для региона 2: от -0.14 до 0.88 млрд рублей
Риск убытков для региона 2: 7.20%
